lib/semantics/resolve-names-utils.cc

```diff
@@ -286,4 +288,356 @@ Bound ArraySpecAnalyzer::GetBound(const parser::SpecificationExpr &x) {
   return Bound{std::move(expr)};
 }
 
+// If SAVE is set on src, set it on all members of dst
+static void PropagateSaveAttr(
+    const EquivalenceObject &src, EquivalenceSet &dst) {
+  if (src.symbol.attrs().test(Attr::SAVE)) {
+    for (auto &obj : dst) {
+      obj.symbol.attrs().set(Attr::SAVE);
+    }
+  }
+}
+static void PropagateSaveAttr(const EquivalenceSet &src, EquivalenceSet &dst) {
+  if (!src.empty()) {
+    PropagateSaveAttr(src.front(), dst);
+  }
+}
+
+void EquivalenceSets::AddToSet(const parser::Designator &designator) {
+  if (CheckDesignator(designator)) {
+    Symbol &symbol{*currObject_.symbol};
+    if (!currSet_.empty()) {
+      // check this symbol against first of set for compatibility
+      Symbol &first{currSet_.front().symbol};
+      CheckCanEquivalence(designator.source, first, symbol) &&
+          CheckCanEquivalence(designator.source, symbol, first);
+    }
+    auto subscripts{currObject_.subscripts};
+    if (subscripts.empty() && symbol.IsObjectArray()) {
+      // record a whole array as its first element
+      for (const ShapeSpec &spec : symbol.get<ObjectEntityDetails>().shape()) {
+        if (auto &lbound{spec.lbound().GetExplicit()}) {
+          if (auto subscript{evaluate::ToInt64(*lbound)}) {
+            subscripts.push_back(*subscript);
+          }
+        }
```

Comment:
else error, unless you've checked elsewhere.

Alternatively, if you run that designator through expression semantics, you could build up a call to the `LBOUND` intrinsic around that expression, fold it, and use the elements of its constant array result.

Reply:
It should have been checked already so I will change it to an assert.